codersdk/deployment.go

```diff
@@ -252,109 +261,1037 @@ type GitAuthConfig struct {
+			YAML:        "gitAuthProviders",
+			Value:       &c.GitAuthProviders,
+		},
+	}
 }
 
 type SupportConfig struct {
-	Links *DeploymentConfigField[[]LinkConfig] `json:"links" typescript:",notnull"`
+	Links bigcli.Object[[]LinkConfig] `json:"links" typescript:",notnull"`
+}
+
+// NewDeploymentConfig returns a new DeploymentConfig without
+// nil values.
+func NewDeploymentConfig() *DeploymentConfig {
+	return &DeploymentConfig{
+		TLS:         &TLSConfig{},
+		Logging:     &LoggingConfig{},
+		Provisioner: &ProvisionerConfig{},
+		RateLimit:   &RateLimitConfig{},
+		Dangerous:   &DangerousConfig{},
+		Trace:       &TraceConfig{},
+		Telemetry:   &TelemetryConfig{},
+		OIDC:        &OIDCConfig{},
+		OAuth2: &OAuth2Config{
+			Github: &OAuth2GithubConfig{},
+		},
+		Pprof:      &PprofConfig{},
+		Prometheus: &PrometheusConfig{},
+		DERP: &DERP{
+			Server: &DERPServerConfig{},
+			Config: &DERPConfig{},
+		},
+		Swagger: &SwaggerConfig{},
+		Support: &SupportConfig{},
+	}
 }
 
 type LinkConfig struct {
-	Name   string `json:"name"`
-	Target string `json:"target"`
-	Icon   string `json:"icon"`
+	Name   string `json:"name" yaml:"name"`
+	Target string `json:"target" yaml:"target"`
+	Icon   string `json:"icon" yaml:"icon"`
 }
 
 type Flaggable interface {
 	string | time.Duration | bool | int | []string | []GitAuthConfig | []LinkConfig
 }
 
-type DeploymentConfigField[T Flaggable] struct {
-	Name  string `json:"name"`
-	Usage string `json:"usage"`
-	Flag  string `json:"flag"`
-	// EnvOverride will override the automatically generated environment
-	// variable name. Useful if you're moving values around but need to keep
-	// backwards compatibility with old environment variable names.
-	//
-	// NOTE: this is not supported for array flags.
-	EnvOverride string `json:"-"`
-	Shorthand   string `json:"shorthand"`
-	Enterprise  bool   `json:"enterprise"`
-	Hidden      bool   `json:"hidden"`
-	Secret      bool   `json:"secret"`
-	Default     T      `json:"default"`
-	Value       T      `json:"value"`
-}
-
-// MarshalJSON removes the Value field from the JSON output of any fields marked Secret.
-// nolint:revive
-func (f *DeploymentConfigField[T]) MarshalJSON() ([]byte, error) {
-	copy := struct {
-		Name       string `json:"name"`
-		Usage      string `json:"usage"`
-		Flag       string `json:"flag"`
-		Shorthand  string `json:"shorthand"`
-		Enterprise bool   `json:"enterprise"`
-		Hidden     bool   `json:"hidden"`
-		Secret     bool   `json:"secret"`
-		Default    T      `json:"default"`
-		Value      T      `json:"value"`
-	}{
-		Name:       f.Name,
-		Usage:      f.Usage,
-		Flag:       f.Flag,
-		Shorthand:  f.Shorthand,
-		Enterprise: f.Enterprise,
-		Hidden:     f.Hidden,
-		Secret:     f.Secret,
-	}
-
-	if !f.Secret {
-		copy.Default = f.Default
-		copy.Value = f.Value
-	}
-
-	return json.Marshal(copy)
+// Scrub returns a copy of the config without secret values.
+func (c *DeploymentConfig) Scrub() (*DeploymentConfig, error) {
```

Comment:
I like the name `Scrub`, but I might name this `WithoutSecrets` instead... it's more verbose but less strenuous to the reader to understand.